### ISSUE TITLE
Connection thrashing observed when min_pool_size is bigger than a particular user's max_user_connections

### ISSUE THREAD

**Issue by waltercacau** [pgbouncer/pgbouncer#647]
Hi there,

We have this use case of a PgBouncer instance that has a few users talking to a single PostgreSQL db and we want to configure different pool sizes for them.

For the purposes of describing the issue, let's say we have two users application and background. We would like to have the pool for (db, application) to keep 30 connections in its pool at all times and for the pool (db, background) to stay with 2 connections at all time. To achieve that, we originally:
- set min_pool_size = 30
- set default_pool_size = 30
- set application's user max_user_connections to 30
- set default max_user_connections to 2

This seems to achieve the desired result but we see constant streams of this in the logs:
```
2021-10-15 18:06:17.994 UTC [7] LOG S-0x55832986dcf0: db/background@address new connection to server (from 172.30.121.106:56060)
2021-10-15 18:06:17.993 UTC [7] LOG S-0x558329870e70: db/background@address closing because: evicted (age=0s)
```

And apparently PG logs indicates that there are many short lived connections (like a close immediately after the connection is created). So this appears to be causing connection thrashing.

After analyzing the code a bit, I believe that this is caused because when the janitor is trying to ensure there are a minimum number of connections in the pool, it does not consider max_user_connections:
https://github.com/pgbouncer/pgbouncer/blob/aace87ef7e7cd9642ed76f2eea1fc0476355b411/src/janitor.c#L490

Could we change the code to also cap pool size by max_user_connections ? My rationale is basically that if a pool has max_user_connections = X, it real pool size can never be more than X. In fact it may practically become lower than X if there are multiple DBs configured with the same user, but at least this prevents connection thrashing in the simple use case of a single DB per user.

Thoughts? Does this seem like a reasonable solution?

Happy to try to contribute a PR, just wanted your opinion on this solution and to make sure I didn't miss anything. Thanks in advance!

**Comment by petere** [comment]
I'm on the fence about this one. If the user says, I want a minimum pool size of X, and I want at most Y connections, and Y <= X, then who says that Y should win? If this is a nonsensical configuration, maybe the user should just fix it? Is there a real use case that would require this kind of configuration to work?

My concern is, PgBouncer by its nature has a lot of settings like this, how many connections, various timeouts, and some of those naturally conflict with each other. If we start addressing each of those conflicts individually, then we might never see the end of it. I'm not opposed to this one in particular, but I wonder if there is a real use for it or if it's just a user mistake.

**Comment by waltercacau** [comment]
That is a fair point.

In terms of a real use case, in my current employer (Instacart) we have found the need for this. For each of our DBs (and their replicas), we wanted to set up a set of bouncers that would be used by multiple users. One user was used by our application (let's call it foreground) and we wanted to maintain a constant N# of server connections to PG so that spikes in traffic didn't cause a sudden spike in new server connections being created (back in 2020 during the initial covid wave we saw drastic traffic spikes and this was one of the measures we used to stabilize our systems during the spikes). To achieve this we set min_pool_size to this user's max_user_connections. However, we also had some other secondary users on these bouncers (e.g. a user for data dog to collect metrics, a user for our a business intelligence tool, ...). Those users had max user connection set to a small value (2). Unfortunately this set up caused connection thrashing and this was the motivation for this issue and the PR I sent.

Let me know if this seems like a reasonable use case.

I could see an alternative feature request to allow setting min pool size on a combination of user + db, but it seemed like a more involved change that would require further discussion.

I think the PR still has some value in avoiding connection thrashing for folks that are either facing the same issue as we did or just unintentionally configured the bouncer in this way.
